### ISSUE TITLE
backspace cannot delete chars sometimes

### ISSUE THREAD

**Issue by FirstLoveLife** [xonsh/xonsh#3314]
```
root@deepin-PC ~/chenli # screenn^H^H^H^H
```
every time I type <kbd>backspace</kbd>, ^H is appended in the end. I have to C-d then re-enter xonsh, and backspace turns into normal.

Where goes wrong?

**Comment by FirstLoveLife** [comment]
UPDATE: konsole doesn't have this problem.

**Comment by scopatz** [comment]
Weird. Thanks for reporting.  What is the output of `xonfig`?

**Comment by shirouto** [comment]
I managed to reproduce this issue consistently.

term: qterminal
xonfig:
+------------------+---------------------+
| xonsh            | 0.9.12              |
| Git SHA          | 50155c66            |
| Commit Date      | Oct 3 15:53:08 2019 |
| Python           | 3.7.3               |
| PLY              | 3.11                |
| have readline    | True                |
| prompt toolkit   | None                |
| shell type       | readline            |
| pygments         | None                |
| on posix         | True                |
| on linux         | True                |
| distro           | ubuntu              |
| on darwin        | False               |
| on windows       | False               |
| on cygwin        | False               |
| on msys2         | False               |                                                                                                                                                                                                     
| is superuser     | False               |                                                                                                                                                                                                     
| default encoding | utf-8               |                                                                                                                                                                                                     
| xonsh encoding   | utf-8               |                                                                                                                                                                                                     
| encoding errors  | surrogateescape     |                                                                                                                                                                                                     
+------------------+---------------------+ 

steps to reproduce:

1. launch ```xonsh``` from ```bash```
2. run ```ls```
3. type in a string like ```fooo``` and then hit backspace thrice which will echo as ```fooo^H^H^H```

I have here basically an off-the-shelf Lubuntu installation without any bash shell customization or terminal tweaking.

**Comment by scopatz** [comment]
> One workaround is to switch to prompt_toolkit.

Thanks for letting us know!  Does this still happen if you have qterminal start xonsh directly?

**Comment by shirouto** [comment]
Yes, it does. ```qterminal -e xonsh --shell-type readline``` will have the same issue described above.

I even turned off the color of the ```ls``` output (since it actually aliases by default to ```ls --color=auto -v``` in this distro ```.bashrc```) and removed layout formatting (```ls -1 | xargs echo```), but the issue remains. However, doing something like ```cat foo.txt``` does not mess the terminal: the backspace still works properly.

**Comment by scopatz** [comment]
Ok, I'll take a look once I am able to get qterminal installed

**Comment by con-f-use** [comment]
`tput smkx` in my `.xonshrc` seems to solve this problem for me, somehow. Although the help says it's supposed to help for <kbd>Del</kbd>

**Comment by anki-code** [comment]
This still actual for xonsh 0.9.26 and could be reproduced by https://github.com/xonsh/xonsh/issues/3314#issuecomment-540149378. The solution from https://github.com/xonsh/xonsh/issues/3314#issuecomment-545145249 is a workaround to fix this.

**Comment by yaxollum** [comment]
Hi, I think this issue is caused by the `restore_tty_sanity` function, because commenting out the last line (`os.system(stty + " sane")`) makes the issue go away.

https://github.com/xonsh/xonsh/blob/5dafceef47aac3db36124d8d91c3905d816685be/xonsh/readline_shell.py#L673-L691

**Comment by anki-code** [comment]
I can't reproduce this on current main (0.23.0) after improving some tty cases.
